diff --git i/config.def.h w/config.def.h
index 877afab..16f7260 100644
--- i/config.def.h
+++ w/config.def.h
@@ -5,7 +5,7 @@
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
  */
-char font[] = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
+char font[] = "monospace-9.6";
 int borderpx = 2;
 
 /*
@@ -134,8 +134,9 @@ unsigned int cursorshape = 2;
  * Default columns and rows numbers
  */
 
-unsigned int cols = 80;
-unsigned int rows = 24;
+// Set geometry to keep it when doFloat
+unsigned int cols = 100;
+unsigned int rows = 30;
 
 /*
  * Default colour and shape of the mouse cursor
@@ -164,12 +165,17 @@ MouseShortcut mshortcuts[] = {
 #define MODKEY Mod1Mask
 #define TERMMOD (ControlMask|ShiftMask)
 
+/* THINK: pass more inner vars from st instance */
+static char *pipe_url[] = { "r.st-pipe", "--url", winid, NULL };
+static char *pipe_vim[] = { "r.st-pipe", "--vim", winid, NULL };
+
 Shortcut shortcuts[] = {
 	/* mask                 keysym          function        argument */
 	{ XK_ANY_MOD,           XK_Break,       sendbreak,      {.i =  0} },
 	{ ControlMask,          XK_Print,       toggleprinter,  {.i =  0} },
 	{ ShiftMask,            XK_Print,       printscreen,    {.i =  0} },
 	{ XK_ANY_MOD,           XK_Print,       printsel,       {.i =  0} },
+	{ TERMMOD,              XK_End,         zoom,           {.f =+10} },
 	{ TERMMOD,              XK_Prior,       zoom,           {.f = +1} },
 	{ TERMMOD,              XK_Next,        zoom,           {.f = -1} },
 	{ TERMMOD,              XK_Home,        zoomreset,      {.f =  0} },
@@ -178,6 +184,9 @@ Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Y,           selpaste,       {.i =  0} },
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
 	{ TERMMOD,              XK_I,           iso14755,       {.i =  0} },
+	{ TERMMOD,              XK_F6,          swapcolors,     {.i =  0} },
+	{ TERMMOD,              XK_U,           externalpipe,   {.v = pipe_url} },
+	{ TERMMOD,              XK_space,       externalpipe,   {.v = pipe_vim} },
 };
 
 /*
