diff --git i/config.def.h w/config.def.h
index 877afab..16f7260 100644
--- i/config.def.h
+++ w/config.def.h
@@ -5,7 +5,7 @@
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
  */
-static char *font = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
+static char *font = "monospace-13";
 static int borderpx = 2;
 
 /*
@@ -176,8 +176,9 @@ unsigned int cursorshape = 2;
  * Default columns and rows numbers
  */
 
-static unsigned int cols = 80;
-static unsigned int rows = 24;
+// Set geometry to keep it when doFloat
+static unsigned int cols = 100;
+static unsigned int rows = 30;
 
 /*
  * Default colour and shape of the mouse cursor
@@ -217,6 +217,10 @@ MouseShortcut mshortcuts[] = {
 #define MODKEY Mod1Mask
 #define TERMMOD (ControlMask|ShiftMask)
 
+/* THINK: pass more inner vars from st instance */
+static char *pipe_url[] = { "r.st-pipe", "--url", NULL };
+static char *pipe_vim[] = { "r.st-pipe", "--vim", NULL };
+
 static Shortcut shortcuts[] = {
	/* mask                 keysym          function        argument */
	{ XK_ANY_MOD,           XK_Break,       sendbreak,      {.i =  0} },
@@ -178,6 +183,10 @@ Shortcut shortcuts[] = {
 	{ TERMMOD,              XK_Y,           selpaste,       {.i =  0} },
 	{ ShiftMask,            XK_Insert,      selpaste,       {.i =  0} },
 	{ TERMMOD,              XK_Num_Lock,    numlock,        {.i =  0} },
+	{ TERMMOD,              XK_F6,          swapcolors,     {.i =  0} },
+	{ TERMMOD,              XK_U,           externalpipe,   {.v = pipe_url} },
+	{ TERMMOD,              XK_space,       externalpipe,   {.v = pipe_vim} },
+	{ TERMMOD,              XK_Return,      newterm,        {.i =  0} },
 };
 
 /*
# @@ -540,2 +540,2 @@ 
#  // Active style
# -#define UNDERCURL_STYLE UNDERCURL_SPIKY
# +#define UNDERCURL_STYLE UNDERCURL_CAPPED
